### ISSUE TITLE
comments-plugin: meta data missing

### ISSUE THREAD

**Issue by pmario** [Jermolene/TiddlyWiki5#3385]
To reproduce:

 - Create a comment for "HelloThere": 
   - creates: `Comment by '(anonymous)' on 'HelloThere'`
 - comment on the comment
   - creates: `Comment by '(anonymous)' on 'Comment by '(anonymous)' on 'HelloThere''`
 - delete the first comment: `Comment by '(anonymous)' on 'HelloThere'`
   - The second comment is gone  <-------- problem 1
 - Create a comment for "HelloThere"
   - creates: `Comment by '(anonymous)' on 'HelloThere'`
   - the second comment magically reappears   <-------- problem 2 
      - That's wrong, because it already lost it's context. 
      - This comment is presented in the wrong context.

.

 - Open `Comment by '(anonymous)' on 'HelloThere'`
 - rename it to: `asdf`
   - The second comment lost it's context and is gone!   <-------- problem 3

. 

 - Level 3 comments are hard be "processed" by humans: `Comment by '(anonymous)' on 'Comment by '(anonymous)' on 'Comment by '(anonymous)' on 'HelloThere'''` 
   - Not talking about level 4 and 5++

# Proposal

## For problem 1 and 2: 

 - There should be a second field eg: `comment-root` that is part of every comment tiddler
  - If the comment is the first comment the `comment-root` should be created
  - If comment-root field exists, the tiddler is a comment already and the field should be copied over to the new comment. 
    - So if a parent, from a "lower level" comment is deleted the comment can still be assigned to the root thread
    - Since `comment-target` points to a tiddler that doesn't exist, this information can be used to show some info, that the comment is "out of context" 
 - Having a `comment-root` field allows to create a [time-line overview like this:](http://web.archive.org/web/20161205105023/http://freestyle-community.tiddlyspace.com/#CommentList)

```
2012-02-28 17:14 - Fb.GKisItaTheme - - - - Fb.GKisItaThemeComment1330445688900 - - - - funkyeah
2012-02-07 15:12 - Fb.GKdivisionThemeGenerator - - - - Fb.GKdivisionThemeGeneratorComment1328623894276 - - - - pmario 
```

## For problem 3: 

 - Deleting a comment, that is referenced by child tiddlers, shouldn't delete the tiddler, but just replace the text-field with: `This comment has been deleted`
 - OR
 - Since "out of context" comments have a `comment-target` and a proposed `comment-root` field we could show an empty parent with `This comment has been deleted`. 

### About the comment title

 - Instead of "cascading" the name for new lower level comments the `comment-root` field could be used
   - eg: `HelloThere/comment/pmario/<date-in-miliseconds>`
   - creation rule: `<comment-root>/comment/<user>/<date-in-miliseconds>`

.

 - This naming schema would be human readable.
 - This naming shema would allow us to use the "Explorer" tab to see all comments related to "HelloThere"
 - <date-in-miliseconds> will allow us to detect "out of context" comments, if a user deletes and recreates a comment. see: problem 2

### About the "listed" comment title

 - At the moment a comment is **shown** like this: `at 08:14:33 Thursday 9th August 2018 `

 - The comment title [could be more like this](http://web.archive.org/web/20161205105023/http://freestyle-community.tiddlyspace.com/#Fb.GKtheAudience): 

```
pmario@Friday, February 3rd, 2012 3:12:53 pm
```

**Comment by AnthonyMuscio** [comment]
Mario,

Just food for thought, perhaps not for the comments plugin but this or other hierarchies, 

This could be considered design algorithms and speculation, it is appropriate here because it extends the proposed algorithm.

Rather than simply store a comment-root why not a comment-branch? You can still delete intermediate comments but reduce title complexity and produce a timeline overview (at any level). 

What if comments were issued an incremental number as their tiddler name and have a user selectable caption applied (and that is displayed in the title) perhaps simply user/date/time (use tiddler subtitle?) and possibly the unique comment number/title (see below). Then using a field called comment-branch in every comment, the first children will contain a comment-branch = [[root-tiddler]] then the grand children comment-branch = [[root-tiddler]] [[Child-comment]] etc... Because we do not expect the title to change for these tiddlers (ever) other than the root tiddler and the "normal" way to list them is from their parent, all should be good. Each "leaf tiddler" simply adds their name at the end of the comment-branch field and use it to create the comment-branch in its children similar to your model. We could have a rename root tiddler tool that would update all children.

This also makes me ask could we copy a comments text, to a child with the first character of every line ">"
>To quote
And comment

This means the structure need not be embedded in the title names which will grow excessively long. Only the  comment-branch would grow and may look like this [[root-tiddler]] 1 2 3  or later [[root-tiddler]] 23 24 25 29

With this you could list all comments for root-tiddler as this is in the comment-branch, or on a search result finding a comment, find the root tiddler and or its children.

It would also be easy to list threads and other views in the future, even if an intermediate tiddler is deleted, however in this case I see value in only archiving comments. One could potentially add a previously posted comment into another branch or "quote" a comment number (where ever it is).

In a more advanced way of managing hierarchies in the future, I plan to name the branch field with a prefix of the root tiddler to allow multiple overlapping hierarchies.

Ideally we would have a generic method to support getting incremental number titles (with or without prefix/suffix) and a generic method to support building branch fields, Which could be used for good and evil as yet unimagined.

Imagine if we had a meta-branch that contained a list of root tiddlers. Or could add ancestors never though of. We could move up and down a branch a bit like your TocP Parent field without ambiguity of multiple tags or  much more.

Am I making any sense?
Tony

**Comment by AnthonyMuscio** [comment]
PS Allowing the use of a prefix on a auto-numbered comment titles will allow them to be system tiddlers and hidden without including system tiddlers.

**Comment by pmario** [comment]
My concern about "branches" and "incremental numbers" is, that copying comments from 1 TW to an other may easily create a valid "tree", where it actually is "out of context". ... 

My concept has the same problem, but I think it is harder using real names instead of just numbers. ...

**Comment by pmario** [comment]
This issue is related to: #3378

**Comment by Jermolene** [comment]
Maybe we can ameliorate the problem a little by adding a timestamp to the titles of comment tiddlers? Or even incorporating a GUID.

I'm not in favour of the "comment-root" field because I think things are less complicated if there's nothing special about a comment vs. a comment root.

**Comment by AnthonyMuscio** [comment]
Just a though while my head is full of "primary keys" and renaming tiddlers. Since we do not want to allow the renaming of comments, What if creating a comment on a tiddler that has no comments created a new and hidden tiddler (even use qualify the first time only), and placed its name in the top tiddler, the first and subsequent comments would all reference this hidden tiddler that will never be renamed and whose name can be found in the root tiddler. Comments on a tiddler will be indicated by reference to the hidden tiddler which is the method by which all comments can be found.

**Comment by Jermolene** [comment]
> Since we do not want to allow the renaming of comments,

That's not correct -- the update I've just made was specifically motivated by the desire to make it possible to rename comments, and tiddlers to which comments are attached.

**Comment by AnthonyMuscio** [comment]
I only saw the need to rename the tiddlers to which they were attached. In the current model you include a higher comments title in the lower comments title. Ie comment titles are auto generated.

In some ways its good to let them retain some integrity.

Regards
Tony

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by pmario** [comment]
>Get Outlook for Android `<spam-link?>`

This link looks like spam to me. please remove it
